[SYCL][Doc] Create  sycl_ext_oneapi_queue_status_query.asciidoc (#5860)
intel/llvm: sycl/doc/extensions/proposed/sycl_ext_oneapi_queue_status_query.asciidoc

Copyright (C) 2022-2022 Intel Corporation.  All rights reserved.

Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks
of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc. used by
permission by Khronos.

== Contact

To report problems with this extension, please open a new issue at:

https://github.com/intel/llvm/issues

== Dependencies

This extension is written against the SYCL 2020 revision 4 specification.  All
references below to the "core SYCL specification" or to section numbers in the
SYCL specification refer to that revision.

== Status

This is a proposed extension specification, intended to gather community
feedback.  Interfaces defined in this specification may not be implemented yet
or may be in a preliminary state.  The specification itself may also change in
incompatible ways before it is finalized.  *Shipping software products should
not rely on APIs defined in this specification.*

== Overview

This extension adds `sycl::queue::size()`, `sycl::queue::empty()`, `sycl::queue::get_wait_list()` member functions.

- `sycl::queue::size()` and `sycl::queue::empty` allow users to query the number of enqueued commands to a queue.

> - `sycl::queue::get_wait_list()` is the queue equivalent of `event::get_wait_list()` 

Reviewer: It's not really necessary (or useful) to list all of the extended member functions here.  They are already described more fully below.  Instead, it's more useful to provide a justification for why this extension exists.  How do you plan to use this extension?  Can you write an overview that starts something like:

> This extension adds new query APIs to the `queue` class which allow a SYCL application to ...

== Specification

=== Feature test macro

This extension provides a feature-test macro as described in the core SYCL
specification.  An implementation supporting this extension must predefine the
macro `SYCL_EXT_ONEAPI_QUEUE_STATUS_QUERY` to one of the values defined in the table
below.  Applications can test for the existence of this macro to determine if
the implementation supports this feature, or applications can test the macro's
value to determine which of the extension's features the implementation
supports.

[%header,cols="1,5"]
|===
|Value
|Description

|1
|Initial version of this extension.
|===

> === Signature and Semantic 

Reviewer 2: ```suggestion
=== Signatures and Semantics 
```

This extension adds the following new member functions to the SYCL `queue` class.
```

> namespace sycl {
> namespace queue {
>  bool empty() const;
>  size_t size() const;
>  std::vector<event> get_wait_list() const;
> }}

Reviewer: ```suggestion
namespace sycl {

class queue {
 bool ext_oneapi_empty() const;
 size_t ext_oneapi_size() const;
 std::vector<event> ext_oneapi_get_wait_list() const;
};

} // namespace sycl
```

These functions are member of `class queue`, not free functions in `namespace queue`.  Also, extended member functions should follow the [extension guidelines](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#_names_for_extensions_to_existing_classes_or_enumerations) and start with the prefix `ext_oneapi_`.

```

The semantics of the new functions are: 

```
empty()
   Returns `true` if `size() == 0`.

> ```

Reviewer: This formatting doesn't work well (see the rendered HTML).  In particular, the backticks aren't interpreted as "code style" format because the whole thing is in a triple-backtick code block.  Normally, we use an AsciiDoc table for function specifications:

```
[frame="topbot",options="header"]
|===
|Function |Description

// --- ROW BREAK ---
a|
[source,c++]
----
ext_oneapi_empty();
----
|
Returns `true` if `size() == 0`.

// --- ROW BREAK ---
a|
[source,c++]
----
ext_oneapi_size();
----
|
Returns the number of enqueued `commands` in the queue that have not been 
completed.  Since the implementation completes commands from the queue
asynchronously, the returned value is a snapshot in time, and the actual 
number of uncompleted commands may be different by the time the function 
returns.  Note that the behavior of `size()` is associated with the SYCL 
`queue` and not with the underlying native-backend-object.

|===
```

```
size_t size() const
   Returns the number of enqueued `commands` in the queue 
   that have not been completed.
   Since the implementation completes commands from the queue 
   asynchronously, the returned value is a snapshot in time,
   and the actual number of uncompleted commands may be different 
   by the time the function returns.
   Note that the behavior of `size()` is associated with the SYCL `queue`
   and not with the underlying native-backend-object.
```

> (based on similar HipSYCL extension https://github.com/illuhad/hipSYCL/blob/develop/doc/extensions.md#hipsycl_ext_queue_wait_list) 

Reviewer: This is not a normative part of the specification, so it doesn't belong in the "Specification" section.  If you want to reference the HipSYCL extension, that could go to the "Overview" section or to some new section at the bottom (that is a peer of "Specification").

// for syntax highlighting purposes.  This is needed because
// docbook uses c++ and html5 uses cpp.
:language: {basebackend@docbook:c++:cpp}

== Notice

Copyright (C) 2022-2022 Intel Corporation.  All rights reserved.

Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks
of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc. used by
permission by Khronos.

== Contact

To report problems with this extension, please open a new issue at:

https://github.com/intel/llvm/issues

== Dependencies

This extension is written against the SYCL 2020 revision 4 specification. All
references below to the "core SYCL specification" or to section numbers in the
SYCL specification refer to that revision.

== Status

This is a proposed extension specification, intended to gather community
feedback. Interfaces defined in this specification may not be implemented yet
or may be in a preliminary state. The specification itself may also change in
incompatible ways before it is finalized.  *Shipping software products should
not rely on APIs defined in this specification.*

== Overview

CUDA has a stream API 

> (link:https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__STREAM.html#group__CUDART__STREAM_1g2021adeb17905c7ec2a3c1bf125c5435[CudaStreamQuery]) with check if all operations in a CUDA stream have been completed. Such feature lack in `sycl::queue`. This extension adds a new query API to introspect queues. In particular, `ext_oneapi_empty` can be used to replace the `CudaStreamQuery`. This should ease the porting applications from CUDA to SYCL.

Reviewer: ```suggestion
(link:https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__STREAM.html#group__CUDART__STREAM_1g2021adeb17905c7ec2a3c1bf125c5435[CudaStreamQuery]) which checks if all operations in a CUDA stream have been completed. Such a feature is lacking in `sycl::queue`. This extension adds a new query API to introspect queues. In particular, `ext_oneapi_empty` can be used to replace the `CudaStreamQuery`. This should ease the porting applications from CUDA to SYCL.
```

== Specification

=== Feature test macro

This extension provides a feature-test macro as described in the core SYCL
specification. An implementation supporting this extension must predefine the
macro `SYCL_EXT_ONEAPI_QUEUE_STATUS_QUERY` to one of the values defined in the table
below. Applications can test for the existence of this macro to determine if
the implementation supports this feature, or applications can test the macro's
value to determine which of the extension's features the implementation
supports.

[%header,cols="2,5"]
|===
|Value
|Description

|1
|Initial version of this extension.
|===

=== Signatures and Semantics 

This extension adds the following new member functions to the SYCL `queue` class.

> ```

Reviewer 3: ```suggestion
[source,c++]
----
```

namespace sycl {

class queue {
 bool ext_oneapi_empty() const;
 size_t ext_oneapi_size() const;
 std::vector<event> ext_oneapi_get_wait_list() const;
};

} // namespace sycl

> ```

Reviewer 3: ```suggestion
----
```

The semantics of the new functions are: 

[frame="topbot",options="header"]
|===
|Function |Description

// --- ROW BREAK ---
a|
[source,c++]
----

> bool ext_oneapi_empty()

Reviewer: ```suggestion
bool ext_oneapi_empty() const
```

----
|

> Returns `true` if `size() == 0`.

Reviewer: ```suggestion
Returns `true` if `ext_oneapi_size() == 0`.
```

// --- ROW BREAK ---
a|
[source,c++]
----

> size_t ext_oneapi_size()

Reviewer: ```suggestion
size_t ext_oneapi_size() const
```

----
|
Returns the number of enqueued `commands` in the queue that have not been 
completed. Since the implementation completes commands from the queue
asynchronously, the returned value is a snapshot in time, and the actual 
number of uncompleted commands may be different by the time the function 

> returns. Note that the behavior of `size()` is associated with the SYCL 

Reviewer: ```suggestion
returns. Note that the behavior of `ext_oneapi_size()` is associated with the SYCL 
```

`queue` and not with the underlying native-backend-object.

// --- ROW BREAK ---
a|
[source,c++]
----
std::vector<event> ext_oneapi_get_wait_list() const
----
|
Returns the list of events such that waiting for all returned events
guarantees that all enqueued commands in the queue have been completed.
Implementations are free to omit events that don't contribute
to the semantic of `get_wait_list()`. For example,
whether already completed events or non-leaf events of the dependency sub-graph
managed by the queue are included in the returned list is implementation-defined.

> This implies that the number of events returned by `get_wait_list()`
> maybe different from the value returned by `size()`.

Reviewer: ```suggestion
This implies that the number of events returned by `ext_oneapi_get_wait_list()`
maybe different from the value returned by `ext_oneapi_size()`.
```

a|
|===

== Similar Work

> - `ext_oneapi_get_wait_list` is inspired by the the HIPSYCL extensions  link:https://github.com/illuhad/hipSYCL/blob/develop/doc/extensions.md#hipsycl_ext_queue_wait_list[queue_wait_list]. 

Reviewer 3: ```suggestion
- `ext_oneapi_get_wait_list` is inspired by the hipSYCL extension link:https://github.com/illuhad/hipSYCL/blob/develop/doc/extensions.md#hipsycl_ext_queue_wait_list[queue_wait_list]. 
```

> 
> 
> 

Reviewer 3: ```suggestion

```
